py/dml/codegen.py

```diff
@@ -1620,58 +1609,55 @@ def try_codegen_invocation(site, init_ast, outargs, location, scope):
             codegen_expression(inarg_ast, location, scope)
             for inarg_ast in inarg_asts]
     if meth_node.fully_typed:
-        return codegen_call(site, meth_node, indices,
-                            inargs, outargs)
+        return codegen_call(site, meth_node, indices, inargs, outargs)
     else:
-        return common_inline(site, meth_node, indices,
-                              inargs, outargs)
+        return common_inline(site, meth_node, indices, inargs, outargs)
 
 @statement_dispatcher
 def stmt_assign(stmt, location, scope):
-    (_, site, tgts_kind, tgt_asts, src_asts) = stmt
-    tgts = [codegen_expression(ast, location, scope) for ast in tgt_asts]
+    (_, site, tgt_ast, src_asts) = stmt
+    assert tgt_ast.kind in ('assign_target_chain', 'assign_target_tuple')
+    tgts = [codegen_expression(ast, location, scope)
+            for ast in tgt_ast.args[0]]
     for tgt in tgts:
         if deep_const(tgt.ctype()):
             raise ECONST(tgt.site)
-
-    if tgts_kind == 'chain':
+    if tgt_ast.kind == 'assign_target_chain':
         method_tgts = [tgts[0]]
-    elif tgts_kind == 'tuple':
-        method_tgts = tgts
     else:
-        raise ICE(site, 'unknown LHS kind')
+        method_tgts = tgts
 
     # TODO support multiple assign sources. It should be generalized.
     method_invocation = try_codegen_invocation(site, src_asts, method_tgts,
                                                location, scope)
     if method_invocation:
```

Comment:
Below this line, there's about 2 lines of code in this function that isn't under a tgt_ast.kind condition. I think control flow will be clearer if you unify those ifs into one, swallowing the cost of duplicating those two lines (avoids flag programming). You might consider breaking out two functions for that. (if this turns out nice, then you could consider one further step: split stmt_assign in two, assign_chain/assign_tuple. Cleaner AST at the cost of a few more lines of dup)

Reply:
Those lines are the following:
```
    lscope = Symtab(scope)
    stmts = []
    ...
    return stmts
```
I'm OK with duplicating `lscope`, but I disagree that duplicating `stmts` and `return stmts` makes the control flow clearer.